### PR TITLE
Solving problem with PH5 library inconsistently modifying sample_rate_multiplier_i when its value is 0 or the column is missing.

### PR DESCRIPTION
**Pull request by damhuonglan**
### What does this PR do?
Solving #386 

### Describe the problem
_pn3 is the previous structure of PH5 that has no column sample_rate_multiplier_i (srm). Before this PR, when reading das_t in experiment.py, srm is consider equal to 1 if it has value 0 or the column is missing. However this is inconsistent with read_arrays() in experiment.py or  query_das_t() in ph5api.py_

### Solution
+ Create tool fix_srm to change srm values from 0 to 1 or add srm=1 in case that column is missing in das tables or array tables. `fix_srm` is a new command. Users need to update the new command to their PH5 api by running
 ```pip install -e .```
 in `ph5` conda environment.
+ In experiment.ReceiversGroup.read_das() and read_arrays(),  and ph5api.PH5.query_das_t(), if srm=0 or missing, error will be thrown recommending user to run fix_srm to fix the value.
+ Add flag -i/--ignore_srm in tabletokef to bypass error thrown for srm=0 or missing when creating kef file. (tabletokef is also used in fix_srm to create backup for the table that are going to be changed, so ignore checking srm is needed for that too.)
+ Add catching experiment.HDF5InteractionError in ph5toms and ph5availability to log as error when function try to read das/array that has srm=0 or missing.

### Checklist
- [ ] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] All tests pass.
- [ ] Any new or changed features have are documented.
- [ ] Changes have been added to `CHANGELOG.txt` .
- [ ] First time contributors have added your name to `CONTRIBUTORS.txt` .

### CODE REVIEW

File: ph5/clients/ph5availability.py [hunk 1]
```diff
@@ -540,15 +551,19 @@ def get_availability(self, station='*', location='*',
...
+                                ph5_das,
+                                chan=channum,
+                                start_epoch=ph5_start_epoch,
+                                stop_epoch=ph5_stop_epoch,
+                                sample_rate=ph5_sample_rate,
+                                sample_rate_multiplier=ph5_multiplier,
+                                check_samplerate=False)
+                        except ph5api.APIError as err:
+                            errors.add(err.msg)
+                            continue
```
Comment (timronan) [inline ph5/clients/ph5availability.py:566]: It seems like the cases where sample rate multiplier are 0 and sample rate multiplier is missing should act the same when it come to ph5availability and ph5toms. Likely, both cases should throw an error message and encourage the user to use `ph5/utilities/fix_srm.py` to fix them. Injecting a temporary number into a field seems confusing and doesn't really fix the problem in the data. 

> Some cases in ph5availability require sample_rate_multiplier_i in calculating, change code to use value 1 if it is missing.

This should probably not be automatic.

File: ph5/core/ph5api.py [hunk 2]
```diff
@@ -800,101 +814,79 @@ def query_das_t(self,
...
-                            'time/ascii_s': row['time/ascii_s'],
-                            'time/epoch_l': row['time/epoch_l'],
-                            'time/micro_seconds_i':
-                                row['time/micro_seconds_i'],
-                            'time/type_s': row['time/type_s'],
-                            'time_table_n_i': row['time_table_n_i']
-                            }
-                das.append(row_dict)
+            if col_srm_exist is True:
+                numexprstr = (
```
Comment (timronan) [inline ph5/core/ph5api.py:826]: What is numexprstr? This seems like a lot of additional refactoring that is occurring in PH5API. Is this directly related to this issue?

File: ph5/utilities/fix_srm.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    group.add_argument("--all_arrays", dest='all_arrays', action='store_true',
+                       default=False,
+                       help=r"Fix sample_rate_multiplier_i in all Array "
+                            "table\n /Experiment_g/Sorts_g/Array_t_xxx .")
+
+    group.add_argument("-A", "--Array_t_", dest="array_t_", metavar="n",
+                       help=r"Fix sample_rate_multiplier_i in Array table"
+                            "\n /Experiment_g/Sorts_g/Array_t_[n].",
+                       type=int)
+
```
Comment (timronan) [inline ph5/utilities/fix_srm.py:58]: I think that only the --all flag should be available. Metadata inconsistencies seems to be one of the biggest issues in PH5 and adding options to create more inconsistencies seems like it could lead to problems in the future. What do you think about this?

**Review by timronan — COMMENTED** [review]
This is a good start, but I am a little bit concerned about creating sample_rate_multiplier inconsistencies between the tables. I think that functions should generally behave consistently when dealing with missing sample rate multipliers and sample rate multipliers set to 0. We need to be careful about injecting fixes into the code but not flagging and updating the underlying data. Ultimately this is a data issue and until the data is all updated the PH5 archive will be inconsistent.

File: create_ph5.sh [hunk 4]
```diff
@@ -36,4 +36,13 @@ cd error
...
 
+cd ../..
+mkdir sampleratemultiplier0
+cd sampleratemultiplier0
+initialize_ph5 -n master.ph5
+keftoph5 -n master.ph5 -k ../../metadata/experiment.kef
+segdtoph5 -n master.ph5 -r ../../segd/1111.0.0.fcnt
+echo "y" | delete_table -n master.ph5 -D 1X1111
+keftoph5 -n master.ph5 -k ../../metadata/Das_t_1X1111.0.0_SRM0.kef
+
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply create_ph5.sh:47]: ph5_no_srm can't be move to `/PH5/ph5/test_data/ph5` because `/PH5/ph5/test_data/ph5` will be deleted whenever TravisCI test is built. I actually start with having ph5_no_srm under that folder, and the result was I had to recreate ph5_no_srm after running create_ph5.sh

I will look at ph5tokef to make sure it run correctly.

File: ph5/clients/ph5availability.py (hunk 1 above)
Comment (damhuonglan) [reply ph5/clients/ph5availability.py:566]: As I stated at the description, in case of sample rate multiplier is missing we will consider its value is 1. The reason is I have no way to add a column to an existing ph5, unless I have to rebuild the whole ph5. I talked about that in the last meeting if you remembered. I know it doesn't really fix the problem in the data. but to handle this problem will be a lot of work. I tried that but was still not successful. We will talk about this again in the meeting today.

File: ph5/core/ph5api.py (hunk 2 above)
Comment (damhuonglan) [reply ph5/core/ph5api.py:826]: Since 
```
            for row in tbl.where(....):
                row_dict = {....}
```
has been repeated a lot and the code took too many lines. I assign the expression inside `tbl.where()` to `numexprstr` so we don't need to repeat that part. It helps developer to maintain the code easier.

File: ph5/utilities/fix_srm.py (hunk 3 above)
Comment (damhuonglan) [reply ph5/utilities/fix_srm.py:58]: In that case, we don't even need --all flag. Just run `fix_srm -n master.ph5 -p [path]`, it will fix the whole ph5 .

File: ph5/clients/ph5availability.py (hunk 1 above)
Comment (timronan) [reply ph5/clients/ph5availability.py:566]: If sample rate multiplier is missing than it seems practical to have the code assume that it is a value of 1. DAS.smr is very difficult to update so this issue needs to be dealt with in software rather than in the data.

File: ph5/utilities/fix_srm.py (hunk 3 above)
Comment (timronan) [reply ph5/utilities/fix_srm.py:58]: Add a verification check so it is not accidentally run.

File: ph5/core/ph5api.py (hunk 2 above)
Comment (timronan) [reply ph5/core/ph5api.py:826]: This a good refactor and should be left in the code. It makes this function more concise.

**Comment by hrotman-pic** [comment]
Commenting here on @timronan's request:
I have a question about the proposed utility for fixing the DAS tables: will it make a copy of the original DAS tables (in case of a mistake)? I think a previous utility to update the DAS tables (fix_3chan_texan) created a copy of the original DAS table for all DAS as well as a copy of the updated DAS table.

(I should add that fix_3chan_texan did not, itself, load the DAS tables it modified.)

File: create_ph5.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (damhuonglan) [reply create_ph5.sh:47]: @timronan I think I should have ph5tokef to create kef file with the exact info from the table in the ph5 file, which means if the table have no 'sample_rate_multiplier', kef file will have no 'sample_rate_multiplier'. Do you agree with that approach?

**Comment by damhuonglan** [comment]
> I have a question about the proposed utility for fixing the DAS tables: will it make a copy of the original DAS tables

@hrotman-pic fix_srm currently creates kef files of the original DAS tables and array tables before fixing them. Is that what you mean by making a copy of original DAS tables?

**Comment by hrotman-pic** [comment]
> > I have a question about the proposed utility for fixing the DAS tables: will it make a copy of the original DAS tables
> 
> @hrotman-pic fix_srm currently creates kef files of the original DAS tables and array tables before fixing them. Is that what you mean by making a copy of original DAS tables?

If the kef files of the original tables exist after the fix, yes, that is what I mean.

File: ph5/entry_points.py [hunk 5]
```diff
@@ -216,6 +216,10 @@ def __init__(self):
                            'ph5.utilities.fix_3chan_texan:main',
                            'For fixing 3-channel Texan data.',
                            type=EntryPointTypes.ALL),
+                EntryPoint('fix_srm',
```
Comment (timronan) [inline ph5/entry_points.py:219]: This command does not seem to be found when running the PH5 from the command prompt. Here is the output:
fix_srm --help
-bash: fix_srm: command not found

**Review by timronan — COMMENTED** [review]
This Pull request is also very close to being done and seems to be working well except and I don't think that fix_srm is working as a command line utility. When I run 
`fix_srm --help or fix_srm -h or fix_srm` I get the error `-bash: fix_srm: command not found`

File: ph5/entry_points.py (hunk 5 above)
Comment (damhuonglan) [reply ph5/entry_points.py:219]: Sorry, this is a new command. You have to update the new command to your PH5 by running `pip install -e .`

Comment (timronan) [reply ph5/entry_points.py:219]: Thanks for the quick feedback. Testing now.

**Review by timronan — APPROVED** [review]
This pull request seems to be working very well and as expected. It seems as though these changes are ready to be merged into the master branch.

**Comment by damhuonglan** [comment]
Thanks. Let me merge master to this branch and update CHANGELOG then request you to re-review it. You can merge it if you want to.

**Review by timronan — COMMENTED** [review]
Thanks for providing the change management log. This PR is ready to be merged.
